models/Document.php

```diff
@@ -695,15 +700,16 @@ public function getChildren($unpublished = false)
      */
     public function hasChildren($unpublished = false)
     {
-        if (is_bool($this->hasChildren)) {
-            if (($this->hasChildren and empty($this->children)) or (!$this->hasChildren and !empty($this->children))) {
-                return $this->getDao()->hasChildren($unpublished);
-            } else {
-                return $this->hasChildren;
+        $cacheKey = $this->getListingCacheKey(func_get_args());
+
+        if (isset($this->hasChildren[$cacheKey])) {
+            if (($this->hasChildren[$cacheKey] and empty($this->children[$cacheKey])) or (!$this->hasChildren[$cacheKey] and !empty($this->children[$cacheKey]))) {
```

Comment:
Is this type of ifs really necessary? If I call this function twice, the database will be queried twice because the children array are still empty. The 2 arguments are protected and  should always be synchronous.

Reply:
I also don't think that this is necessary, changed that, see 0d44a8a3074057a561f42cd527684e6f90445fad